fix: documentation around diff-aware scanning (#661)
returntocorp/semgrep-docs: docs/semgrep-ci/configuration-reference.md

## Diff-aware scanning (`SEMGREP_BASELINE_REF`)

> There are two ways to run a semgrep scan: as a full scan or as a diff-aware scan.

Reviewer: ```suggestion
There are two ways to run a semgrep scan:
```

> Semgrep scans can be classified by scope. The scope of a scan refers to what lines of code are scanned in a codebase. When classifying scans by scope, there are two types of scans:

Reviewer: @adamkvitek This is the only line I've changed since our call since you brought up the ambiguity of "types of scan". Let me know if "scope" is better.

Reviewer 2: LGTM!

> A full scan will scan your entire code base and report every finding in the code base. We recommend doing a full scan of your `main` branch at a regular cadence, such as every night or every week, depending on your own needs. This will ensure that Semgrep App has a full list of all findings in your code base, regardless of when they were introduced. To run a full scan, simply run `semgrep ci` without setting `SEMGREP_BASELINE_REF` environment variabel.

Reviewer: ```suggestion
<dl><dt>Full scan</dt>
<dd>A full scan will scan your entire codebase and report every finding in the codebase. It is recommended to perform a full scan of your `main` branch at a regular cadence, such as every night or every week, depending on your own needs. This ensures that Semgrep App has a full list of all findings in your code base, regardless of when they were introduced. To run a full scan, simply run `semgrep ci` without setting `SEMGREP_BASELINE_REF` environment variable.
```

For @chmccreery : I just converted the paragraphs into a "description list" (<dl> tag).

    <dt>Full scan</dt>

>     <dd>A full scan will scan your entire codebase and report every finding in the codebase. It is recommended to perform a full scan of your <code>main</code> branch at a regular cadence, such as every night or every week, depending on your own needs. This ensures that Semgrep App has a full list of all findings in your code base, regardless of when they were introduced. To run a full scan, simply run <code>semgrep ci</code> without setting <code>SEMGREP_BASELINE_REF</code> environment variable.</dd>

Reviewer 2: - Do not use will in technical writing. Only use will when you are referring to the time that needs to elapse between action and goal. See https://developers.google.com/style/tense?hl=en.
- Since I removed `will`, I am trying to avoid `scan scans` in the first sentence.

```suggestion
    <dd><p>A full scan is scanning your entire codebase and reports every finding in the codebase. It is recommended to perform a full scan of your <code>main</code> branch at a regular cadence, such as every night or every week. This ensures that Semgrep App has a full list of all findings in your code base, regardless of when they were introduced. To run a full scan, run <code>semgrep ci</code> without setting <code>SEMGREP_BASELINE_REF</code> environment variable.</p></dd>
```

Reviewer: Honestly, "is scanning" is not better than "will scan".

I think will is fine because this is not a procedure. If the user chooses to use a full scan, then they will scan their codebase. Does that make sense?

Reviewer 2: One of the reasons why we do this is that some engineers are using automatic translations of pages and some languages do not have future tense (so will cause a bit of havoc). @s-santillan See Google style guide suggestion "However, the fact that the reader will be writing and running code in the future isn't a good reason to use future tense. Stick with present tense where possible, as shown in the previous Recommended example." From my experience, TWs are only allowed to use "will" really when you are talking about things such as "The hedgehog will hibernate starting in October." (actual specific future). It might be ok in this case, but I would avoid it here also. I will leave this decision up to you. (Please stick with the division to paragraphs I suggest.)

Reviewer: I think I'll go with "A full scan runs on your entire codebase..."

> A diff-aware scan will scan your code before and after some "baseline" and only report findings that are newly introduced in the commits after that baseline. For example, imagine you have a repository with 10 commits and you set commit number 8 as the baseline. Consequently, Semgrep only returns scan results introduced by changes in commits 9 and 10. This is how `semgrep ci` should be run in pull requests and merge requests, since it will report only the findings that are created by those code changes. To run a diff-aware scan, use `SEMGREP_BASELINE_REF=<ref> semgrep ci` where ref can be a commit hash, branch name, or other git ref.

Reviewer: ```suggestion
<dt>Diff-aware scan</dt>
<dd>A diff-aware scan runs on your code before and after some "baseline" and only reports findings that are newly introduced in the commits after that baseline. For example, imagine you have a repository with 10 commits and you set commit number 8 as the baseline. Consequently, Semgrep only returns scan results introduced by changes in commits 9 and 10. This is how `semgrep ci` should be run in pull requests and merge requests, since it will report only the findings that are created by those code changes. To run a diff-aware scan, use `SEMGREP_BASELINE_REF=<ref> semgrep ci` where ref can be a commit hash, branch name, or other git ref.
</dd></dl>
```

> :::note

Author: They are actually two separate notes. One about only doing full scans for the main branch, and one about never doing full scans on anything but the main branch. They are related, but feel like two different misconfigurations with different repercussions which is why I separated them out. Thoughts?

> :::note

Reviewer: I merged the notes into a list. It was a little hard to set this as a suggestion (because I was also fixing the tags) so I committed it. Resolving this now.

>     <dd>A diff-aware scan runs on your code before and after some "baseline" and only reports findings that are newly introduced in the commits after that baseline. For example, imagine you have a repository with 10 commits and you set commit number 8 as the baseline. Consequently, Semgrep only returns scan results introduced by changes in commits 9 and 10. This is how <code>semgrep ci</code> should be run in pull requests and merge requests, since it will report only the findings that are created by those code changes. To run a diff-aware scan, use <code>SEMGREP_BASELINE_REF=<span class="placeholder">REF</span> semgrep ci</code> where <span class="placeholder">REF</span> can be a commit hash, branch name, or other git ref.</dd>

Reviewer 2: - @chmccreery Ad "should" - `This is how <code>semgrep ci</code> should be run in pull requests and merge requests` - do we want to say: 
A. This is recommended setup for Semgrep in CI.
B. This is how Semgrep should work in CI, but it often does not (uncertainty implied by modal words such as "should").
Avoid these words and rather specify. See https://developers.google.com/style/word-list?hl=en#should.
- @s-santillan I am not sure if we can use `semgrep ci` - some time ago we discussed this with @pabloest that we should rather state Semgrep in CI. I am not sure if there was any additional discussion made that I was not part of regarding this topic.
- git refs are probably [git references](https://git-scm.com/book/en/v2/Git-Internals-Git-References).

```suggestion
    <dd><p>A diff-aware scan runs on your code before and after some "baseline" and only reports findings that are newly introduced in the commits after that baseline.</p>
    <p>For example, imagine a hypothetical repository with 10 commits. You set commit number 8 as the baseline. Consequently, Semgrep only returns scan results introduced by changes in commits 9 and 10. This is how Semgrep in CI can run in pull requests and merge requests, since it reports only the findings that are created by those code changes. To run a diff-aware scan, use <code>SEMGREP_BASELINE_REF=<span class="placeholder">REF</span> semgrep ci</code> where <span class="placeholder">REF</span> can be a commit hash, branch name, or other git reference.</p></dd>
```

Reviewer 2: Another note - try to build this, if the output is satisfactory.

Reviewer: @adamkvitek Semgrep CI in this context is literally a command, `semgrep ci`.

Reviewer: I'm accepting the suggestion but reverting "Semgrep in CI" to `semgrep ci`.

</dl>

:::note

> * Do not perform diff-aware scans on your `main` branch. The Semgrep App keeps track of which findings have been fixed on a given branch. If you configure diff scans on your main branch, and compare the last commit to the penultimate commit, Semgrep wrongly considers all findings from before the penultimate commit to be fixed.

Reviewer 2: ```suggestion
* Do **not** perform diff-aware scans on your `main` branch. The Semgrep App keeps track of which findings have been fixed on a given branch. If you configure diff scans on your main branch, and compare the last commit to the penultimate commit, Semgrep wrongly considers all findings from before the penultimate commit to be fixed.
```

> * Do not perform full scans on non-mainline or non-trunk branches. Performing full scans on every branch will slow down your CI jobs, show developers findings they didn't introduce, and result in many duplicated findings in the Semgrep App, resulting in poorer experience there as well.

Reviewer 2: - I assume that `show developers findings they didn't introduce` means `displays findings that developers did not introduce` - the meaning was a bit confusing.
- Do not use words such as "show", use display instead when possible.
- Make the not bold when you want to strongly discourage from doing something (otherwise try to frame it positively).

```suggestion
* Do **not** perform full scans on non-mainline or non-trunk branches. Performing full scans on every branch slows down your CI jobs, displays findings that developers did not introduce, and results in many duplicated findings in the Semgrep App, resulting in a poorer experience.
```

:::
 
### Examples of `SEMGREP_BASELINE_REF`

To only report findings newly added
since branching off from your `main` branch, set the following:

> <pre class="language-bash"><code>SEMGREP_BASELINE_REF=<span className="placeholder">main</span></code></pre>

Reviewer 2: Consider making the placeholder upper case as per https://developers.google.com/style/placeholders.
